Add TencentARC PhotoMaker support

bssrdf opened the pull request:
Hi, this is an implementation of TencentARC [PhotoMaker](https://github.com/TencentARC/PhotoMaker). I am putting it in draft mode for now to get feedbacks.
|          |         |
|----    |----   |
| <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/7a9b55dc-10e5-449b-96ad-62b141cccd87" alt="image" width="512" height="auto"> | <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/00b430e2-b703-4a72-87d2-055d3f698338" alt="image" width="512" height="auto"> |
| <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/fd3f91b5-8d32-4c45-8df1-dfb33401d05f" alt="image" width="512" height="auto"> | <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/d0994d18-2be6-4fec-a113-7634df10291e" alt="image" width="512" height="auto"> |
| <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/208aa21a-2689-4741-8ac6-60e25119284f" alt="image" width="512" height="auto"> | <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/4c083f55-2d91-477b-8a22-dfc14ee894b6" alt="image" width="512" height="auto"> |
| <img src="https://github.com/TencentARC/PhotoMaker/assets/689043/0cd219a6-9e9d-480f-ae02-76292f9cfd56" alt="image" width="512" height="auto"> | <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/6fb5d8ad-16df-4e1d-934a-7f68d08c4881" alt="image" width="512" height="auto"> |
| <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/c2f0f319-8f1e-4eac-815a-e76b2b52ec31" alt="image" width="512" height="auto"> | <img src="https://github.com/leejet/stable-diffusion.cpp/assets/689043/b10b03d1-39fe-4d3a-b6f6-979089524329" alt="image" width="512" height="auto"> |


~The results are not quite as good as the official version and I have to turn on vae tiling because of 8GB GPU mem limit and vae tiling has seam issues visible.~

ID fidelity has been improved and an option to offload VAE latent decode to cpu was added. Now quality is much better.

Note: this requires an upstream [fix](https://github.com/ggerganov/ggml/pull/737) in ```GGML```.

Amin456789 commented:
yes yes yes! thank u so much for ur hard work, with this and upcoming inpainting and svd in future sd cpp will be the best sd of all!

leejet commented:
Great job! Could you merge the changes in the master branch? I just merged a PR(https://github.com/leejet/stable-diffusion.cpp/pull/159) that introduces GGMLBlock, making it more convenient to implement neural networks using ggml.

bssrdf commented:
> Great job! Could you merge the changes in the master branch? I just merged a PR(#159) that introduces GGMLBlock, making it more convenient to implement neural networks using ggml.

@leejet, sounds good. Now I need to merge with master and resolve conflicts first. 

Note: to test the functionality, ```ggml``` must be updated to the latest which contains some bug fixes and missing/updated functions. I will also need to put a safetensor version of the photomaker model file I converted somewhere, maybe on huggingface.

bssrdf commented:
I really wished I could have done this development after updating to your merge (those are really some fundamental changes)!  @leejet , are you sure the newly added CLIPVision model is working? The implementation of photomaker highly depends on it.

leejet commented:
> Note: to test the functionality, ggml must be updated to the latest which contains some bug fixes and missing/updated functions.

I have updated the ggml in the master branch to the latest commit. To adapt to the latest commit, I had to make some changes.

> are you sure the newly added CLIPVision model is working? The implementation of photomaker highly depends on it.

I tested with the weights of clip vision from the stable video diffusion, and in some cases, I encountered NaN issues. This seems to be related to certain operators within ggml. I have now updated ggml to the latest commit, but I'm not sure if this issue has been resolved. I'll verify it later when I have some free time. I believe the overall network structure implementation in clip vision is sound because my clip vision implementation reuses most of the clip text structure, which has been validated and found to be correct.

bssrdf commented:
@leejet, I finally made it. Now PhotoMaker is fully working so I pull this PR out from draft. 

I had to make some changes (add two new classes) to CLIP and MultiHeadAttention for my implementation and fixed a couple bugs along the way. Maybe we can resolve the differences during teh review.

Anyway,  I don't want to wait too long to deal with another big merge. The bottom line is it didn't break other modes.

README.md is also updated with instructions so anyone can try it.

FSSRepo commented:
@bssrdf `DGGML_CUDA_FORCE_MMQ` Is this option strictly necessary?

bssrdf commented:
> @bssrdf `DGGML_CUDA_FORCE_MMQ` Is this option strictly necessary?
> 
> `pmid.h` has a lot of commented code.

OOps, that option is only for older GPUs without tensor cores. I have removed it and also cleaned up comments in pmid.hpp.
Thanks for reviewing.

leejet commented:
Great! I'll find time in the next few days to review your changes.

Green-Sky commented on clip.hpp:
```diff
@@ -530,29 +530,23 @@ struct CLIPLayer : public GGMLBlock {
+        blocks["self_attn"]   = std::shared_ptr<GGMLBlock>(new MultiheadAttention(d_model, n_head, true, atten1));
+
         blocks["layer_norm1"] = std::shared_ptr<GGMLBlock>(new LayerNorm(d_model));
         blocks["layer_norm2"] = std::shared_ptr<GGMLBlock>(new LayerNorm(d_model));
```
saw this while skimming.
you generally want to use `std::make_shared<LayerNorm>(d_model)` instead, it offers a bunch of small optimizations and clean up the code. (it implicitly casts to `std::shared_ptr<GGMLBlock>`)

bssrdf replied:
Fair point, but these constructs are not from this PR and they already exist in many other places. Maybe a separate PR can address them. Thanks for reviewing.

Green-Sky replied:
yea, should be looked at after this merges.

leejet commented:
I have reviewed the entire architectural changes and made some improvements. Everything is okay, but there are some code portions that I think can be reused. I will finish the modifications shortly, and then we can merge. Thank you for your amazing work.

Green-Sky commented on stable-diffusion.h:
```diff
@@ -145,7 +145,7 @@ SD_API sd_image_t* txt2img(sd_ctx_t* sd_ctx,
                            float style_strength,
                            bool normalize_input,
-                           std::vector<sd_image_t*> &input_id_images);
+                           const char* input_id_images_path);
```
personally not a fan of using paths here instead of images.
But I do think we should provide a convenience function that loads an image from path, since that is an common use case (but not always the case).

leejet replied:
Yes, for the API, passing image data in memory is more elegant than directly passing a file path. I will take some time later to explore how to improve that.

bssrdf commented:
> I have reviewed the entire architectural changes and made some improvements. Everything is okay, but there are some code portions that I think can be reused. I will finish the modifications shortly, and then we can merge. Thank you for your amazing work.

Sounds good. Thanks for the code reusing updates. Let me know if you find CLIPVisionModel can be customized/shared so CLIPVisionModel2 can be removed.

leejet commented:
@bssrdf I have completed the code reuse for the CLIPVisionModel and LoraModel. I have tested it on my machine, and everything is working well. Could you please take some time to test it as well? If there are no issues, I will proceed to merge this pull request.

bssrdf commented:
> @bssrdf I have completed the code reuse for the CLIPVisionModel and LoraModel. I have tested it on my machine, and everything is working well. Could you please take some time to test it as well? If there are no issues, I will proceed to merge this pull request.

Thanks, @leejet. Appreciate your time consolidating CLIPVisionModel and LoraModel. 
My first try of running PhotoMaker crashed with segfault and cuda OOM during building PM's lora model  : 
```
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: NVIDIA GeForce GTX 1070, compute capability 6.1, VMM: yes
[INFO ] stable-diffusion.cpp:158  - loading model from '../models/RealVisXL_V3.0.safetensors'
[INFO ] model.cpp:705  - load ../models/RealVisXL_V3.0.safetensors using safetensors format
[INFO ] stable-diffusion.cpp:181  - Stable Diffusion XL
[INFO ] stable-diffusion.cpp:187  - Stable Diffusion weight type: f16
[WARN ] stable-diffusion.cpp:193  - !!!It looks like you are using SDXL model. If you find that the generated images are completely black, try specifying SDXL VAE FP16 Fix with the --vae parameter. You can find it here: https://huggingface.co/madebyollin/sdxl-vae-fp16-fix/blob/main/sdxl_vae.safetensors
[INFO ] stable-diffusion.cpp:233  - VAE Autoencoder: Using CPU backend
[INFO ] model.cpp:705  - load ../models/photomaker-v1.safetensors using safetensors format
[INFO ] lora.hpp:38   - loading LoRA from '../models/photomaker-v1.safetensors'
ggml_backend_cuda_buffer_type_alloc_buffer: allocating 354.38 MiB on device 0: cudaMalloc failed: out of memory
[ERROR] ggml_extend.hpp:873  - lora alloc params backend buffer failed
[ERROR] model.cpp:1390 - read tensor data failed: '../models/photomaker-v1.safetensors'
Segmentation fault
```

leejet commented:
@bssrdf I tried your latest commit 16c9da0ccadcb0be94ee1dad7e48dc5abf5fba81 and my latest commit df28af9c70aa685d7ecac62f72788b9f30f5b030. Looking at the logs, there doesn't seem to be much difference in VRAM usage between the two. Could it be that there were other processes on your system using a significant amount of VRAM at that time and causing insufficient availability?

16c9da0ccadcb0be94ee1dad7e48dc5abf5fba81

```
[INFO ] stable-diffusion.cpp:157  - loading model from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
[INFO ] model.cpp:676  - load ..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors using safetensors format
[DEBUG] model.cpp:742  - init from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
[INFO ] stable-diffusion.cpp:168  - loading vae from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
[INFO ] model.cpp:676  - load ..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors using safetensors format
[DEBUG] model.cpp:742  - init from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
[INFO ] stable-diffusion.cpp:181  - Stable Diffusion XL
[INFO ] stable-diffusion.cpp:187  - Stable Diffusion weight type: f16
[DEBUG] stable-diffusion.cpp:188  - ggml tensor size = 432 bytes
[DEBUG] ggml_extend.hpp:881  - clip params backend buffer size =  1564.36 MB (713 tensors)
[DEBUG] ggml_extend.hpp:881  - unet params backend buffer size =  4900.07 MB (1680 tensors)
[DEBUG] ggml_extend.hpp:881  - vae params backend buffer size =  94.47 MB (140 tensors)
[INFO ] model.cpp:676  - load ..\models\photomaker-v1.safetensors using safetensors format
[DEBUG] model.cpp:742  - init from '..\models\photomaker-v1.safetensors'
[INFO ] pmid.hpp:405  - loading LoRA from '..\models\photomaker-v1.safetensors'
[DEBUG] model.cpp:1315 - loading tensors from ..\models\photomaker-v1.safetensors
[DEBUG] ggml_extend.hpp:881  - lora_pmid params backend buffer size =  354.38 MB (10240 tensors)
[DEBUG] model.cpp:1315 - loading tensors from ..\models\photomaker-v1.safetensors
[DEBUG] pmid.hpp:450  - finished loaded lora
[INFO ] stable-diffusion.cpp:266  - loading stacked ID embedding (PHOTOMAKER) model file from '..\models\photomaker-v1.safetensors'
[INFO ] model.cpp:676  - load ..\models\photomaker-v1.safetensors using safetensors format
[DEBUG] model.cpp:742  - init from '..\models\photomaker-v1.safetensors'
[DEBUG] ggml_extend.hpp:881  - pmid params backend buffer size =  1243.48 MB (407 tensors)
```

df28af9c70aa685d7ecac62f72788b9f30f5b030
```
[INFO ] stable-diffusion.cpp:158  - loading model from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
[INFO ] model.cpp:705  - load ..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors using safetensors format
[DEBUG] model.cpp:771  - init from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
[INFO ] stable-diffusion.cpp:169  - loading vae from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
[INFO ] model.cpp:705  - load ..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors using safetensors format
[DEBUG] model.cpp:771  - init from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
[INFO ] stable-diffusion.cpp:181  - Stable Diffusion XL
[INFO ] stable-diffusion.cpp:187  - Stable Diffusion weight type: f16
[DEBUG] stable-diffusion.cpp:188  - ggml tensor size = 432 bytes
[DEBUG] ggml_extend.hpp:878  - clip params backend buffer size =  1564.36 MB (713 tensors)
[DEBUG] ggml_extend.hpp:878  - unet params backend buffer size =  4900.07 MB (1680 tensors)
[DEBUG] ggml_extend.hpp:878  - vae params backend buffer size =  94.47 MB (140 tensors)
[INFO ] model.cpp:705  - load ..\models\photomaker-v1.safetensors using safetensors format
[DEBUG] model.cpp:771  - init from '..\models\photomaker-v1.safetensors'
[INFO ] lora.hpp:38   - loading LoRA from '..\models\photomaker-v1.safetensors'
[DEBUG] model.cpp:1343 - loading tensors from ..\models\photomaker-v1.safetensors
[DEBUG] ggml_extend.hpp:878  - lora params backend buffer size =  354.38 MB (10240 tensors)
[DEBUG] model.cpp:1343 - loading tensors from ..\models\photomaker-v1.safetensors
[DEBUG] lora.hpp:74   - finished loaded lora
[INFO ] stable-diffusion.cpp:264  - loading stacked ID embedding (PHOTOMAKER) model file from '..\models\photomaker-v1.safetensors'
[INFO ] model.cpp:705  - load ..\models\photomaker-v1.safetensors using safetensors format
[DEBUG] model.cpp:771  - init from '..\models\photomaker-v1.safetensors'
[DEBUG] ggml_extend.hpp:878  - pmid params backend buffer size =  1243.48 MB (407 tensors)
```

leejet commented:
I added a parameter, `--clip-on-cpu`, to run CLIP on the CPU. This should reduce VRAM usage by ~2GB for PhotoMaker.
no `--clilp-on-cpu`
```sh
[INFO ] stable-diffusion.cpp:414  - total params memory size = 7182.38MB (VRAM 7182.38MB, RAM 0.00MB): clip 1564.36MB(VRAM), unet 4900.07MB(VRAM), vae 94.47MB(VRAM), controlnet 0.00MB(VRAM), pmid 623.48MB(VRAM)
```

`--clilp-on-cpu`
```sh
[INFO ] stable-diffusion.cpp:414  - total params memory size = 7182.38MB (VRAM 4994.54MB, RAM 2187.84MB): clip 1564.36MB(RAM), unet 4900.07MB(VRAM), vae 94.47MB(VRAM), controlnet 0.00MB(VRAM), pmid 623.48MB(RAM)
```

bssrdf commented:
> @bssrdf I tried your latest commit [16c9da0](https://github.com/leejet/stable-diffusion.cpp/commit/16c9da0ccadcb0be94ee1dad7e48dc5abf5fba81) and my latest commit [df28af9](https://github.com/leejet/stable-diffusion.cpp/commit/df28af9c70aa685d7ecac62f72788b9f30f5b030). Looking at the logs, there doesn't seem to be much difference in VRAM usage between the two. Could it be that there were other processes on your system using a significant amount of VRAM at that time and causing insufficient availability?
> 
> [16c9da0](https://github.com/leejet/stable-diffusion.cpp/commit/16c9da0ccadcb0be94ee1dad7e48dc5abf5fba81)
> 
> ```
> [INFO ] stable-diffusion.cpp:157  - loading model from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
> [INFO ] model.cpp:676  - load ..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors using safetensors format
> [DEBUG] model.cpp:742  - init from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
> [INFO ] stable-diffusion.cpp:168  - loading vae from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
> [INFO ] model.cpp:676  - load ..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors using safetensors format
> [DEBUG] model.cpp:742  - init from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
> [INFO ] stable-diffusion.cpp:181  - Stable Diffusion XL
> [INFO ] stable-diffusion.cpp:187  - Stable Diffusion weight type: f16
> [DEBUG] stable-diffusion.cpp:188  - ggml tensor size = 432 bytes
> [DEBUG] ggml_extend.hpp:881  - clip params backend buffer size =  1564.36 MB (713 tensors)
> [DEBUG] ggml_extend.hpp:881  - unet params backend buffer size =  4900.07 MB (1680 tensors)
> [DEBUG] ggml_extend.hpp:881  - vae params backend buffer size =  94.47 MB (140 tensors)
> [INFO ] model.cpp:676  - load ..\models\photomaker-v1.safetensors using safetensors format
> [DEBUG] model.cpp:742  - init from '..\models\photomaker-v1.safetensors'
> [INFO ] pmid.hpp:405  - loading LoRA from '..\models\photomaker-v1.safetensors'
> [DEBUG] model.cpp:1315 - loading tensors from ..\models\photomaker-v1.safetensors
> [DEBUG] ggml_extend.hpp:881  - lora_pmid params backend buffer size =  354.38 MB (10240 tensors)
> [DEBUG] model.cpp:1315 - loading tensors from ..\models\photomaker-v1.safetensors
> [DEBUG] pmid.hpp:450  - finished loaded lora
> [INFO ] stable-diffusion.cpp:266  - loading stacked ID embedding (PHOTOMAKER) model file from '..\models\photomaker-v1.safetensors'
> [INFO ] model.cpp:676  - load ..\models\photomaker-v1.safetensors using safetensors format
> [DEBUG] model.cpp:742  - init from '..\models\photomaker-v1.safetensors'
> [DEBUG] ggml_extend.hpp:881  - pmid params backend buffer size =  1243.48 MB (407 tensors)
> ```
> 
> [df28af9](https://github.com/leejet/stable-diffusion.cpp/commit/df28af9c70aa685d7ecac62f72788b9f30f5b030)
> 
> ```
> [INFO ] stable-diffusion.cpp:158  - loading model from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
> [INFO ] model.cpp:705  - load ..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors using safetensors format
> [DEBUG] model.cpp:771  - init from '..\..\stable-diffusion-webui\models\Stable-diffusion\sd_xl_base_1.0.safetensors'
> [INFO ] stable-diffusion.cpp:169  - loading vae from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
> [INFO ] model.cpp:705  - load ..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors using safetensors format
> [DEBUG] model.cpp:771  - init from '..\..\stable-diffusion-webui\models\VAE\sdxl_vae-fp16-fix.safetensors'
> [INFO ] stable-diffusion.cpp:181  - Stable Diffusion XL
> [INFO ] stable-diffusion.cpp:187  - Stable Diffusion weight type: f16
> [DEBUG] stable-diffusion.cpp:188  - ggml tensor size = 432 bytes
> [DEBUG] ggml_extend.hpp:878  - clip params backend buffer size =  1564.36 MB (713 tensors)
> [DEBUG] ggml_extend.hpp:878  - unet params backend buffer size =  4900.07 MB (1680 tensors)
> [DEBUG] ggml_extend.hpp:878  - vae params backend buffer size =  94.47 MB (140 tensors)
> [INFO ] model.cpp:705  - load ..\models\photomaker-v1.safetensors using safetensors format
> [DEBUG] model.cpp:771  - init from '..\models\photomaker-v1.safetensors'
> [INFO ] lora.hpp:38   - loading LoRA from '..\models\photomaker-v1.safetensors'
> [DEBUG] model.cpp:1343 - loading tensors from ..\models\photomaker-v1.safetensors
> [DEBUG] ggml_extend.hpp:878  - lora params backend buffer size =  354.38 MB (10240 tensors)
> [DEBUG] model.cpp:1343 - loading tensors from ..\models\photomaker-v1.safetensors
> [DEBUG] lora.hpp:74   - finished loaded lora
> [INFO ] stable-diffusion.cpp:264  - loading stacked ID embedding (PHOTOMAKER) model file from '..\models\photomaker-v1.safetensors'
> [INFO ] model.cpp:705  - load ..\models\photomaker-v1.safetensors using safetensors format
> [DEBUG] model.cpp:771  - init from '..\models\photomaker-v1.safetensors'
> [DEBUG] ggml_extend.hpp:878  - pmid params backend buffer size =  1243.48 MB (407 tensors)
> ```

@leejet, yes, the segfault and OOM error are caused by my WSL instance. Restarting it fixed the issue. My test of Photomaker runs ok. I think it is ready for the final merge. Many thanks.

leejet commented:
Great! Thank you for your contribution.
